File: 1544.make-the-string-great.py
#
# @lc app=leetcode id=1544 lang=python3
#
# [1544] Make The String Great
#

# @lc code=start
class Solution:
    def makeGood(self, s: str) -> str:

        stack = []
        for c in s:
            #the ASCII diff between upper and lower case is 32
            if stack and abs(ord(stack[-1]) - ord(c)) == 32:
                stack.pop()
            else:
                stack.append(c)
        return "".join(stack)
        
# @lc code=end

File: 1544.make-the-string-great.java
/*
 * @lc app=leetcode id=1544 lang=java
 *
 * [1544] Make The String Great
 */

// @lc code=start
import java.util.*;
/*
clarify: Remove paris of same letter where lower case == upper case. After removal, the rest of string will concate.
brutal: use while loop, each time use a window of size 2 to find all the pairs to remove until we cannot find any pairs.
bottleneck: After removing a pair, we do not know whether the left and right string will lead to another removal.
keyword: adjacent removal -> stack
optimize:use a stack to store all the valid characters, pop when removal happens between the top of the stack and the next char in the string.

*/
class Solution {
    public String makeGood(String s) {
        Deque<Character> stack = new ArrayDeque<>();
        for(char c : s.toCharArray()){
            if(!stack.isEmpty() && Math.abs(stack.peek() - c) == 32){
                stack.pop();
            }else{//stack is empty or the removal does not happen
                stack.push(c);
            }
        }
        StringBuilder ans = new StringBuilder();
        while(!stack.isEmpty()){
            ans.append(stack.pop());
        }
        return ans.reverse().toString();
    }
}
// @lc code=end

